rust/operator-binary/src/druid_controller.rs

```diff
@@ -0,0 +1,632 @@
+        rolegroup: RoleGroupRef,
+    },
+    #[snafu(display("object {} is missing metadata to build owner reference", druid))]
+    ObjectMissingMetadataForOwnerRef {
+        source: stackable_operator::error::Error,
+        druid: ObjectRef<DruidCluster>,
+    },
+    #[snafu(display("failed to update status of {}", druid))]
+    ApplyStatus {
+        source: stackable_operator::error::Error,
+        druid: ObjectRef<DruidCluster>,
+    },
+    #[snafu(display(
+        "Failed to get ZooKeeper connection string from config map {} in namespace {}",
+        cm_name,
+        namespace
+    ))]
+    GetZookeeperConnStringConfigMap {
+        source: stackable_operator::error::Error,
+        cm_name: String,
+        namespace: String,
+    },
+    #[snafu(display(
+        "Failed to get ZooKeeper connection string from config map {} in namespace {}",
+        cm_name,
+        namespace
+    ))]
+    MissingZookeeperConnString { cm_name: String, namespace: String },
+}
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub async fn reconcile_druid(druid: DruidCluster, ctx: Context<Ctx>) -> Result<ReconcilerAction> {
+    tracing::info!("Starting reconcile");
+    let druid_ref = ObjectRef::from_obj(&druid);
+    let client = &ctx.get_ref().client;
+
+    let zk_confmap = druid.spec.zookeeper_reference.config_map_name.clone();
+    let zk_cm_ns = druid.spec.zookeeper_reference.namespace.clone();
+    let zk_connstr = client
+        .get::<ConfigMap>(&zk_confmap, Some(&zk_cm_ns)) // TODO
+        .await
+        .with_context(|| GetZookeeperConnStringConfigMap {
+            cm_name: zk_confmap.clone(), // TODO
+            namespace: zk_cm_ns.clone(), // TODO
+        })?
+        .data
+        .and_then(|mut data| data.remove("ZOOKEEPER"))
+        .with_context(|| MissingZookeeperConnString {
+            cm_name: zk_confmap.clone(), // TODO
+            namespace: zk_cm_ns.clone(), // TODO
+        })?;
+
+    let mut roles = HashMap::new();
+
+    let config_files = vec![
+        PropertyNameKind::Env,
+        PropertyNameKind::File(JVM_CONFIG.to_string()),
+        PropertyNameKind::File(LOG4J2_CONFIG.to_string()),
+        PropertyNameKind::File(RUNTIME_PROPS.to_string()),
+    ];
+
+    roles.insert(
+        DruidRole::Broker.to_string(),
+        (config_files.clone(), druid.spec.brokers.clone()),
+    );
+
+    roles.insert(
+        DruidRole::Coordinator.to_string(),
+        (config_files.clone(), druid.spec.coordinators.clone()),
+    );
+
+    roles.insert(
+        DruidRole::Historical.to_string(),
+        (config_files.clone(), druid.spec.historicals.clone()),
+    );
+
+    roles.insert(
+        DruidRole::MiddleManager.to_string(),
+        (config_files.clone(), druid.spec.middle_managers.clone()),
+    );
+
+    roles.insert(
+        DruidRole::Router.to_string(),
+        (config_files, druid.spec.routers.clone()),
+    );
```

Comment:
Copy&Paste where `DruidCluster::get_role` could be used.

Reply:
https://github.com/stackabletech/druid-operator/pull/59/commits/e60228e977488133a8983449fccfcfb1f3ce1fa6